op-service/sources/blob_client.go

```diff
@@ -0,0 +1,103 @@
+package sources
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto/kzg4844"
+)
+
+type BSCBlobClient struct {
+	// BSCBlobClient will rotate client.RPC in pool whenever a client runs into an error while fetching blobs
+	pool *ClientPool[client.RPC]
+}
+
+func NewBSCBlobClient(clients []client.RPC) *BSCBlobClient {
+	return &BSCBlobClient{
+		pool: NewClientPool[client.RPC](clients...),
+	}
+}
+
+func (s *BSCBlobClient) GetBlobs(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.Blob, error) {
+	if len(hashes) == 0 {
+		return []*eth.Blob{}, nil
+	}
+
+	blobSidecars, err := s.GetBlobSidecars(ctx, ref)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob sidecars for L1BlockRef %s: %w", ref, err)
+	}
+
+	validatedBlobs, err := validateBlobSidecars(blobSidecars, ref)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate blob sidecars for L1BlockRef %s: %w", ref, err)
+	}
+
+	blobs := make([]*eth.Blob, len(hashes))
+	for i, indexedBlobHash := range hashes {
+		blob, ok := validatedBlobs[indexedBlobHash.Hash]
+		if !ok {
+			return nil, fmt.Errorf("blob sidecars fetched from rpc mismatched with expected hash %s for L1BlockRef %s", indexedBlobHash.Hash, ref)
+		}
+		blobs[i] = blob
+	}
+	return blobs, nil
+}
+
+func (s *BSCBlobClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRef) (eth.BSCBlobSidecars, error) {
+	var errs []error
+	for i := 0; i < s.pool.Len(); i++ {
+		var blobSidecars eth.BSCBlobSidecars
+
+		f := s.pool.Get()
```

Comment:
It would be better to ensure that the pool frequently evicts (or fixes) false clients, otherwise, the latency might increase rapidly when quite a number of the clients fail.